common/cuda_hip/distributed/matrix_kernels.hpp.inc

```diff
@@ -0,0 +1,290 @@
+            auto col_part = col_part_ids[thrust::get<1>(tuple)];
+            return row_part == local_part && col_part != local_part;
+        });
+
+    // 1. sort global columns, part-id and range-id according to
+    // their part-id and global columns
+    // the previous `non_local_global_col_idxs` is not modify to
+    // keep it consistent with the non-local row and values array
+    array<GlobalIndexType> sorted_non_local_global_col_idxs{
+        exec, non_local_global_col_idxs};
+    auto key_it = thrust::make_zip_iterator(
+        thrust::make_tuple(non_local_col_part_ids.get_data(),
+                           sorted_non_local_global_col_idxs.get_data()));
+    thrust::sort_by_key(thrust::device, key_it, key_it + num_non_local_elements,
+                        non_local_col_range_ids.get_data());
+
+    // 2. remove duplicate columns, now the new column i has global index
+    // non_local_global_col_idxs[i]
+    auto non_local_global_col_idxs_begin =
+        sorted_non_local_global_col_idxs.get_data();
+    auto non_local_global_col_idxs_end = thrust::get<0>(thrust::unique_by_key(
+        thrust::device, non_local_global_col_idxs_begin,
+        non_local_global_col_idxs_begin + num_non_local_elements,
+        thrust::make_zip_iterator(
+            thrust::make_tuple(non_local_col_part_ids.get_data(),
+                               non_local_col_range_ids.get_data()))));
+    auto num_non_local_cols = static_cast<size_type>(thrust::distance(
+        non_local_global_col_idxs_begin, non_local_global_col_idxs_end));
+
+    // 2.5 copy unique_columns to non_local_to_global map
+    non_local_to_global.resize_and_reset(num_non_local_cols);
+    exec->copy(num_non_local_cols, non_local_global_col_idxs_begin,
+               non_local_to_global.get_data());
+
+    // 3. create mapping from unique_columns
+    // since we don't have hash tables on GPUs I'm first sorting the non-local
+    // global column indices and their new local index again by the global
+    // column index. Then I'm using binary searches to find the new local column
+    // index.
+    array<LocalIndexType> permutation{exec, num_non_local_cols};
+    thrust::sequence(thrust::device, permutation.get_data(),
+                     permutation.get_data() + num_non_local_cols);
+    thrust::sort_by_key(
+        thrust::device, non_local_global_col_idxs_begin,
+        non_local_global_col_idxs_begin + num_non_local_cols,
+        thrust::make_zip_iterator(thrust::make_tuple(
+            non_local_col_part_ids.get_data(), permutation.get_data())));
+
+    // 4. map column index of non-local entries to new columns
+    non_local_col_idxs.resize_and_reset(num_non_local_elements);
+    array<size_type> lower_bounds{exec, num_non_local_elements};
+    // I have to precompute the lower bounds because the calling binary
+    // searches from the device does not work:
+    // https://github.com/NVIDIA/thrust/issues/1415
```

Comment:
Maybe add a TODO here so that we know to fix it once it has been fixed from thrust's side ?

Reply:
Yeah, I think that is okay. A TODO tag just allows us to keep track of possible improvements. :)

Reply:
yes, that would make us very future-proof ;)

Reply:
As far as I understood the issue I've linked to, they plan to deprecate the device-side launch of lower bounds anyway, so it probably will never be possible to do. But it seems very unclear how they actually want to proceed with it, so I will add the TODO.

Reply:
you can also use our own binary search kernels in searching.cuh/hpp.inc. But I would like to do some preprocessing for this operation anyways :)